Allow long status code

devlead opened the pull request:
Relates to and fixes issue #1135

dnfclas commented:
Hi __@devlead__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

mathewc commented on src/WebJobs.Script/Binding/HttpBinding.cs:
```diff
@@ -124,7 +124,7 @@ internal static void ParseResponseObject(IDictionary<string, object> responseObj
                 if ((responseObject.TryGetValue("statusCode", out statusValue, ignoreCase: true) ||
                      responseObject.TryGetValue("status", out statusValue, ignoreCase: true)) &&
-                     (statusValue is int || statusValue is string))
+                     (statusValue is long || statusValue is int || statusValue is string))
```
For completeness we should probably handle all the integral types, e.g. int, long, ulong, short, ushort, ...

Recommend adding an `internal static bool TryParseStatusCode(object value)` method that is used here. Internally it can then do all the right checks for the rest of the integral types. Also please add a unit test to HttpBindingTests.cs for the new helper. Can be a [Theory] test that tests the various inputs, e.g. like [this test](https://github.com/Azure/azure-webjobs-sdk-script/blob/da6e7ed5d606bd7a2c70e2b3f262220852abfd39/test/WebJobs.Script.Tests/FeatureFlagsTests.cs#L17).

devlead replied:
Sure I'll address that 👍

mamaso commented on src/WebJobs.Script/Binding/HttpBinding.cs:
```diff
@@ -137,11 +132,54 @@ internal static void ParseResponseObject(IDictionary<string, object> responseObj
+                statusValue is uint ||
+                statusValue is ushort ||
+                statusValue is byte)
+            {
```
I don't think we want to support byte or sbyte as they don't cover the full range of status codes.  Also, can we cast directly to `HttpStatusCode` without Convert.ToInt32(, allowing us to move this code up to the int case above?

mathewc replied:
I'm merging his PR now, and I'll make that minor change.

devlead replied:
Not done yet :wink:
(int)(object)200l wont work but (int)(long)(object)200l will so can do that instead of convert.

Sure I can remove  byte and sbyte.

fabiocav replied:
Provided a bit of feedback to @mathewc on how we can avoid dealing with all the different types and simplify this a little bit.

devlead commented:
@mathewc PR rebased & updated
* Extracted logic to internal static TryParseStatusCode 
* Added support for Integral Types long, short, ulong, uint, ushort and HttpStatusCode
* Added test ParseResponseObject_ObjectStatus_ReturnsExpectedResult
```csharp 
        [Theory]
        [InlineData((short)301, HttpStatusCode.MovedPermanently)]
        [InlineData((ushort)401, HttpStatusCode.Unauthorized)]
        [InlineData((int)501, HttpStatusCode.NotImplemented)]
        [InlineData((uint)202, HttpStatusCode.Accepted)]
        [InlineData((long)302, HttpStatusCode.Redirect)]
        [InlineData((ulong)402, HttpStatusCode.PaymentRequired)]
        [InlineData(HttpStatusCode.Conflict, HttpStatusCode.Conflict)]
        [InlineData("410", HttpStatusCode.Gone)]
        [InlineData("", HttpStatusCode.OK)]
```
result
```
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: 301, expected: MovedPermanently)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: 401, expected: Unauthorized)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: 501, expected: NotImplemented)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: 202, expected: Accepted)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: 302, expected: Redirect)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: 402, expected: PaymentRequired)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: Conflict, expected: Conflict)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: "410", expected: Gone)
[00:01:27] Passed   Microsoft.Azure.WebJobs.Script.Tests.HttpBindingTests.ParseResponseObject_ObjectStatus_ReturnsExpectedResult(value: "", expected: OK)
```

fabiocav commented:
Awesome! As mentioned on the previous review comment, I think we can compress and simplify the code a bit here. Mathew is merging your PR and making that small tweak.

devlead commented on src/WebJobs.Script/Binding/HttpBinding.cs:
```diff
@@ -137,11 +136,80 @@ internal static void ParseResponseObject(IDictionary<string, object> responseObj
+            var stringValue = statusValue as string;
+            if (stringValue == null)
+            {
+                return true;
```
This should false, should I fix & rebase or are you in middle of merging?

mathewc replied:
Fixed

devlead commented on test/WebJobs.Script.Tests/Binding/HttpBindingTests.cs:
```diff
@@ -212,5 +212,30 @@ public void ParseResponseObject_StatusWithNullBody_ReturnsExpectedResult()
+        [InlineData((ulong)402, HttpStatusCode.PaymentRequired)]
+        [InlineData(HttpStatusCode.Conflict, HttpStatusCode.Conflict)]
+        [InlineData("410", HttpStatusCode.Gone)]
+        [InlineData("", HttpStatusCode.OK)]
```
Should add null test.

devlead commented:
Soon going to bed here, but if your in a hurry you've got commit rights to my branch 😉 or I could submit new PR later.

mathewc commented:
Merged, thanks :)

devlead commented:
Excellent :shipit:
